protocols/ping/src/handler.rs

```diff
@@ -21,107 +21,207 @@
+    /// Configuration options.
+    config: PingConfig,
+    /// The timer for when to send the next ping.
     next_ping: Delay,
+    /// The connection timeout.
+    connection_timeout: Duration,
+    /// The current keep-alive, i.e. until when the connection that
+    /// the handler operates on should be kept alive.
+    connection_keepalive: KeepAlive,
+    /// The last result from an inbound or outbound ping.
+    last_result: Option<PingResult>,
+    _marker: std::marker::PhantomData<TSubstream>
 }
 
 impl<TSubstream> PingHandler<TSubstream>
 where
     TSubstream: AsyncRead + AsyncWrite,
 {
-    /// Builds a new `PingHandler`.
-    pub fn new() -> Self {
-        // TODO: allow customizing timeout; depends on https://github.com/libp2p/rust-libp2p/issues/864
+    /// Builds a new `PingHandler` with the given configuration.
+    pub fn new(config: PingConfig) -> Self {
+        let now = Instant::now();
+        let connection_timeout = config.timeout * config.max_timeouts.get();
+        let connection_keepalive = KeepAlive::Until(now + connection_timeout);
+        let next_ping = Delay::new(now);
         PingHandler {
-            inner: OneShotHandler::default(),
-            next_ping: Delay::new(Instant::now()),
-            delay_to_next_ping: Duration::from_secs(15),
+            config,
+            next_ping,
+            connection_timeout,
+            connection_keepalive,
+            last_result: None,
+            _marker: std::marker::PhantomData
         }
     }
 }
 
-impl<TSubstream> Default for PingHandler<TSubstream>
-where
-    TSubstream: AsyncRead + AsyncWrite,
-{
-    #[inline]
-    fn default() -> Self {
-        PingHandler::new()
-    }
-}
-
 impl<TSubstream> ProtocolsHandler for PingHandler<TSubstream>
 where
     TSubstream: AsyncRead + AsyncWrite,
 {
-    type InEvent = void::Void;
-    type OutEvent = protocol::PingOutput;
+    type InEvent = Void;
+    type OutEvent = PingResult;
     type Error = ProtocolsHandlerUpgrErr<io::Error>;
     type Substream = TSubstream;
     type InboundProtocol = protocol::Ping;
     type OutboundProtocol = protocol::Ping;
     type OutboundOpenInfo = ();
 
-    fn listen_protocol(&self) -> Self::InboundProtocol {
-        self.inner.listen_protocol()
+    /// The outbound ping timeout, as specified in the [`PingConfig`].
+    fn outbound_timeout(&self) -> Duration {
+        self.config.timeout
     }
 
-    fn inject_fully_negotiated_inbound(&mut self, protocol: ()) {
-        self.inner.inject_fully_negotiated_inbound(protocol)
+    fn listen_protocol(&self) -> protocol::Ping {
+        protocol::Ping
     }
 
-    fn inject_fully_negotiated_outbound(&mut self, duration: Duration, info: Self::OutboundOpenInfo) {
-        self.inner.inject_fully_negotiated_outbound(duration, info)
+    fn inject_fully_negotiated_inbound(&mut self, _: ()) {
+        // A ping from a remote peer has been answered.
+        self.last_result = Some(Ok(PingSuccess::Pong));
```

Comment:
If we receive both an inbound and an outbound ping simultaneously, one of them will be ignored because they both write over `last_result`.

At the moment I think that we call `poll()` after every single method call, but I can imagine a future where we call multiple methods other than `poll` in a row, then only `poll` at the end.

Alternatively, we could also enforce that `poll()` gets called after every single method call, and write a doc-comment about this in the `ProtocolsHandler` trait definition.

Reply:
I am aware of the possible overriding of the `last_result` and considered queuing all results but I currently don't think there is actually any harm in only keeping the  `last_result`, do you? I mean, it is clear that the precision of the ping-based keep-alive managed by this handler is determined by how frequently it is being polled and if multiple inbound/outbound substream outputs are "injected" between polling, it just means that this precision suffers. Do you think that a) it would be documented more clearly that only the result of the last inbound or outbound substream upgrade is remembered between invocations of `poll` or b) that all results should be queued? I don't have a strong opinion on the matter, the current implementation just seemed simpler and I didn't see much harm in overriding the occasional intermediate result.

Reply:
I'm also unsure of what to do.